docs/specs/Plugins_spec.md

```diff
@@ -0,0 +1,162 @@
+# Feature Proposal: Plugins
+
+## General
+
+This spec describes plugin system that will make Stitch more flexible.
+
+These are the domains that can be extended by using plugins.
+
+1. Graphql Schema extension
+2. Apollo Server extension
+3. Fastify extension
+4. Stitch Concepts Extensions
+
+## Plugin Interface
+
+Each Stitch plugin should include the subset of the following properties:
+
+```yaml
+name: My plugin
+spec:
+  graphql:
+    typeDefs: ITypeDefinitions
+    resolvers: IResolvers<any, TContext> | Array<IResolvers<any, TContext>>
+    directiveResolvers: IDirectiveResolvers
+    schemaDirectives: Record<string, typeof SchemaDirectiveVisitor>
+    schemaDirectivesContext: Record<string, unknown>
+  apollo:
+    plugins: ApolloServerPlugin[]
+  fastify:
+    middlewares: FastifyMiddleware[]
+    decorators: FastifyDecorator[]
+    plugins: FastifyPlugin[]
+    hooks: FastifyHook[]
+  stitch:
+    argument_injection_builtIns: any[]
+    upstreams: Upstream[]
+```
+
+## API Reference
+
+---
+
+### GraphQL
+
+Plugins can extend GraphQL engine of Stitch by added custom directives, scalars or types with resolvers. Despite declaration-only schema enhancement that can be done using Registry this feature should be used when there is some new functionality that should be added to Stitch Graphql engine.
+
+#### typeDefs
+
+The definition of custom directives and/or scalars and types.
+as defined [here](https://github.com/ardatan/graphql-tools/blob/e366a09e31760e6dc9f14239f953fa52a0ed53df/src/Interfaces.ts#L89)
+
+---
+
+#### resolvers
+
+Custom scalars or types resolvers as defined [here](https://github.com/ardatan/graphql-tools/blob/e366a09e31760e6dc9f14239f953fa52a0ed53df/src/Interfaces.ts#L97)
+
+---
+
+#### directiveResolvers
+
+> The custom directives can be implemented either using resolver function or defining class that inherits from SchemaDirectiveVisitor.
+
+This property contains custom directives resolvers.
+
+---
+
+#### schemaDirectives
+
+Custom directives inherit SchemaDirectiveVisitor
+
+---
+
+#### schemaDirectivesContext
+
+The context sent to `SchemaDirectiveVisitor.visitSchemaDirectives` method
+
+---
+
+### Apollo Server
+
+Apollo server plugins allows to add some event hooks. For example, metrics measurement or logging. Also, [base policy](../authorization.md) concept is implemented in Stitch using Apollo Server plugin.
+
+#### Apollo Server Plugins
```

Comment:
I think the approach you suggest is more elegant. The only thing that should be solved before is to unite two places the Apollo Server configuration is built.
1. Building executable schema https://github.com/Soluto/stitch/blob/4fc839a1792e969203ae63ea8d6ace6553e5cdd6/services/src/modules/graphql-service.ts#L75
2. Actually creating server config https://github.com/Soluto/stitch/blob/4fc839a1792e969203ae63ea8d6ace6553e5cdd6/services/src/modules/gateway.ts#L18

After it will be done we can apply config transformations there.